alica_engine/include/engine/blackboard/Blackboard.h

```diff
@@ -1,47 +1,325 @@
+                    _vals.at(targetKey) = makeBBValueForIndex<false>::make(_vals.at(targetKey).index(), std::forward<decltype(srcValue)>(srcValue));
+                },
+                srcBb._vals.at(srcKey));
+    }
+
+    bool hasValue(const std::string& key) const { return _vals.count(key); }
+    void removeValue(const std::string& key) { _vals.erase(key); }
+
+    void clear() { _vals.clear(); }
+    bool empty() const { return _vals.empty(); }
+    size_t size() const { return _vals.size(); }
+
+private:
+    std::unordered_map<std::string, BBValueType> _vals;
+    std::unordered_map<std::string, std::string> _yamlType;
+
+    friend BlackboardUtil;
+
+    template <class T>
+    static const char* getTypeName()
+    {
+        auto typeIndex = BBValueType{T{}}.index() - 1;
+        if (typeIndex < BB_VALUE_TYPE_NAMES_SIZE) {
+            // known type (after conversions to supported types in the the variant are taken into account)
+            return BB_VALUE_TYPE_NAMES[typeIndex];
+        }
+        // Note: std::any is also considered to be an unknown type
+        return nullptr;
     }
-    std::any& get(const std::string& key) { return vals.at(key); }
-    const std::any& get(const std::string& key) const { return vals.at(key); }
 
-    void set(const std::string& key, const std::any& value) { vals.at(key) = value; }
+    static const char* getTypeName(std::size_t index)
+    {
+        if (index - 1 < BB_VALUE_TYPE_NAMES_SIZE) {
+            // index - 1 to account for std::monostate in the variant
+            return BB_VALUE_TYPE_NAMES[index - 1];
+        }
+        // Note: if index points to std::any, nullptr will be returned, because std::any is also considered to be an unknown type
+        return nullptr;
+    }
+
+    static std::optional<std::size_t> getTypeIndex(const std::string& type)
+    {
+        for (std::size_t i = 0; i < BB_VALUE_TYPE_NAMES_SIZE; ++i) {
+            if (BB_VALUE_TYPE_NAMES[i] == type) {
+                // i + 1 to account for std::monostate in the variant
+                return i + 1;
+            }
+        }
+        return {};
+    }
+
+    static constexpr const char* BB_VALUE_TYPE_NAMES[] = {"bool", "int64", "double", "std::string"};
+    static constexpr std::size_t BB_VALUE_TYPE_NAMES_SIZE = sizeof(BB_VALUE_TYPE_NAMES) / sizeof(const char*);
+
+    template <bool PARSE_ARGS = false>
+    struct makeBBValueForIndex
+    {
+        template <class... Args>
+        static BBValueType make(std::size_t index, Args&&... args)
+        {
+            // make a BBValueType constructed from a value of type that is at `index` of the variant
+            // The value itself is either constructed from args or parsed from args
+            // Note: index cannot be std::monostate
+            // throws if variant construction fails
+            assert(index > 0);
+            return makeHelper(index, std::make_index_sequence<BB_VALUE_TYPE_NAMES_SIZE + 2>(), std::forward<Args>(args)...);
```

Comment:
Can you add a comment why + 2 is used here? I think it was because we have to iterate over the 2 additional types in std::variant, monostate and any, right?

Reply:
yes